### PR TITLE
Fix race in the implementation of __tsan_acquire()

### PR DESCRIPTION
**Pull request by daveclausen**
`__tsan::Acquire()`, which is called by `__tsan_acquire()`, has a performance optimization which attempts to avoid acquiring the atomic variable's mutex if the variable has no associated memory model state.

However, if the atomic variable was recently written to by a `compare_exchange_weak/strong` on another thread, the memory model state may be created *after* the atomic variable is updated. This is a data race, and can cause the thread calling `Acquire()` to not realize that the atomic variable was previously written to by another thread.

Specifically, if you have code that writes to an atomic variable using `compare_exchange_weak/strong`, and then in another thread you read the value using a relaxed load, followed by an `atomic_thread_fence(memory_order_acquire)`, followed by a call to `__tsan_acquire()`, TSAN may not realize that the store happened before the fence, and so it will complain about any other variables you access from both threads if the thread-safety of those accesses depended on the happens-before relationship between the store and the fence.

This change eliminates the unsafe optimization in `Acquire()`. Now, `Acquire()` acquires the mutex before checking for the existence of the memory model state.

### CODE REVIEW

File: compiler-rt/test/tsan/compare_exchange_acquire_fence.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// RUN: %clangxx_tsan -O1 %s -o %t && %run %t 2>&1 | FileCheck %s
```
Comment (vitalybuka) [inline compiler-rt/test/tsan/compare_exchange_acquire_fence.cpp:1]: Exit code 0 should be enough
we probably do not need  | FileCheck here

**Review by vitalybuka — APPROVED** [review]
Can you run ninja`LIT_OPTS="--time-tests" ninja ... `
We may need to tune test if it in the top.

File: compiler-rt/test/tsan/compare_exchange_acquire_fence.cpp (hunk 1 above)
Comment (daveclausen) [reply compiler-rt/test/tsan/compare_exchange_acquire_fence.cpp:1]: Done.

I also removed the CHECK and CHECK-NOT comment lines at the end of the file.

**Review by daveclausen — COMMENTED** [review]
It doesn't show up in the list of slowest tests:

```
$ LIT_OPTS="--time-tests" ninja check-tsan

...

Slowest Tests:
--------------------------------------------------------------------------
45.32s: ThreadSanitizer-x86_64 :: restore_stack.cpp
33.56s: ThreadSanitizer-x86_64 :: Linux/check_memcpy.c
21.56s: ThreadSanitizer-x86_64 :: bench_threads.cpp
14.70s: ThreadSanitizer-x86_64 :: deadlock_detector_stress_test.cpp
14.65s: ThreadSanitizer-x86_64 :: signal_thread.cpp
12.60s: ThreadSanitizer-x86_64 :: stress.cpp
10.81s: ThreadSanitizer-x86_64 :: Linux/fork_multithreaded4.cpp
9.79s: ThreadSanitizer-x86_64 :: force_background_thread.cpp
9.53s: ThreadSanitizer-x86_64 :: lots_of_threads.c
7.14s: ThreadSanitizer-x86_64 :: deep_stack1.cpp
5.73s: ThreadSanitizer-x86_64 :: fd_close_norace3.cpp
5.40s: ThreadSanitizer-x86_64 :: signal_reset.cpp
5.19s: ThreadSanitizer-x86_64 :: mmap_stress.cpp
4.84s: ThreadSanitizer-x86_64 :: signal_sync2.cpp
4.48s: ThreadSanitizer-x86_64 :: global_race.cpp
4.41s: ThreadSanitizer-x86_64 :: ignore_lib6.cpp
3.65s: ThreadSanitizer-x86_64 :: mmap_stress2.cpp
3.51s: ThreadSanitizer-x86_64 :: must_deadlock.cpp
3.51s: ThreadSanitizer-x86_64 :: simple_stack.c
3.50s: ThreadSanitizer-x86_64 :: setuid2.c

Tests Times:
--------------------------------------------------------------------------
[    Range    ] :: [               Percentage               ] :: [ Count ]
--------------------------------------------------------------------------
[44.0s,46.0s) :: [                                        ] :: [  1/409]
[42.0s,44.0s) :: [                                        ] :: [  0/409]
[40.0s,42.0s) :: [                                        ] :: [  0/409]
[38.0s,40.0s) :: [                                        ] :: [  0/409]
[36.0s,38.0s) :: [                                        ] :: [  0/409]
[34.0s,36.0s) :: [                                        ] :: [  1/409]
[32.0s,34.0s) :: [                                        ] :: [  0/409]
[30.0s,32.0s) :: [                                        ] :: [  0/409]
[28.0s,30.0s) :: [                                        ] :: [  0/409]
[26.0s,28.0s) :: [                                        ] :: [  0/409]
[24.0s,26.0s) :: [                                        ] :: [  0/409]
[22.0s,24.0s) :: [                                        ] :: [  0/409]
[20.0s,22.0s) :: [                                        ] :: [  1/409]
[18.0s,20.0s) :: [                                        ] :: [  0/409]
[16.0s,18.0s) :: [                                        ] :: [  0/409]
[14.0s,16.0s) :: [                                        ] :: [  2/409]
[12.0s,14.0s) :: [                                        ] :: [  1/409]
[10.0s,12.0s) :: [                                        ] :: [  1/409]
[ 8.0s,10.0s) :: [                                        ] :: [  2/409]
[ 6.0s, 8.0s) :: [                                        ] :: [  1/409]
[ 4.0s, 6.0s) :: [                                        ] :: [  6/409]
[ 2.0s, 4.0s) :: [*                                       ] :: [ 16/409]
[ 0.0s, 2.0s) :: [************************************    ] :: [377/409]
--------------------------------------------------------------------------

Testing Time: 45.37s

Total Discovered Tests: 451
  Unsupported      :  88 (19.51%)
  Passed           : 362 (80.27%)
  Expectedly Failed:   1 (0.22%)
```

**Comment by vitalybuka** [comment]
Merging having that @dvyukov LGTMed code in internal conversation.

**Comment by vitalybuka** [comment]
It needs %link_libcxx_tsan, like in llvm-project/compiler-rt/test/tsan/compare_exchange.cpp

**Pull request merged**